django/thunderstore/community/models/package_listing.py

```diff
@@ -378,6 +380,82 @@ def get_has_perms() -> bool:
+        if self.visibility.public_detail:
+            return True
+
+        if user is None:
+            return False
+
+        if self.visibility.owner_detail:
+            if self.package.owner.can_user_access(user):
+                return True
+
+        if self.visibility.moderator_detail:
+            for listing in self.package.community_listings.all():
+                if listing.community.can_user_manage_packages(user):
+                    return True
+
+        if self.visibility.admin_detail:
+            if user.is_superuser:
+                return True
+
+        return False
+
+    @transaction.atomic
+    def update_visibility(self):
+        """
+        Updates the package listing's visibility based on whether its package is active, its review status,
+        and the visibility of its active versions.
+
+        By default, listings are visible to everyone (for now). Rejected listings aren't publicly visible,
+        and listings with inactive packages aren't visible at all.
+        """
+        original_visibility_bitstring = self.visibility.bitstring()
+
+        self.visibility.public_detail = True
+        self.visibility.public_list = True
+        self.visibility.owner_detail = True
+        self.visibility.owner_list = True
+        self.visibility.moderator_detail = True
+        self.visibility.moderator_list = True
+
+        if not self.package.is_active:
+            self.visibility.public_detail = False
+            self.visibility.public_list = False
+            self.visibility.owner_detail = False
+            self.visibility.owner_list = False
+            self.visibility.moderator_detail = False
+            self.visibility.moderator_list = False
+
+        if self.review_status == PackageListingReviewStatus.rejected:
+            self.visibility.public_detail = False
+            self.visibility.public_list = False
+
+        if (
+            self.community.require_package_listing_approval
+            and self.review_status == PackageListingReviewStatus.unreviewed
+        ):
+            self.visibility.public_detail = False
+            self.visibility.public_list = False
+
+        versions = self.package.versions.filter(is_active=True).all()
+        if versions.exclude(visibility__public_detail=False).count() == 0:
+            self.visibility.public_detail = False
+        if versions.exclude(visibility__public_list=False).count() == 0:
+            self.visibility.public_list = False
+        if versions.exclude(visibility__owner_detail=False).count() == 0:
+            self.visibility.owner_detail = False
+        if versions.exclude(visibility__owner_list=False).count() == 0:
+            self.visibility.owner_list = False
+        if versions.exclude(visibility__moderator_detail=False).count() == 0:
+            self.visibility.moderator_detail = False
+        if versions.exclude(visibility__moderator_list=False).count() == 0:
+            self.visibility.moderator_list = False
```

Comment:
This runs a dozen of queries, could easily be reduced to a single query and filtering logic done on the python level

Reply:
Also I think the dependency on a package's versions is sort of odd here, shouldn't the listing track whether the package itself is visible? So the version based visibility condition updates should probably be moved to the package instance, and the listing should track the package's visibility + its own conditions for review status

Reply:
I think it's more a matter of when than if. Not doing it in this PR is likely useful for limiting the scope of this specific PR, but perhaps you could implement the function of "get what the visibility should be for this package" already as standalone and utilize the output of that here, and then later on in a followup PR make use of it for the Package's visibility fields update?